File: 3/3.cpp
#include "3.h"

#include <iostream>
#include <memory>
#include <optional>
#include <string>
#include <vector>

#include "filing.h"

const size_t START_LENGTH = std::string("mul(").length();
const size_t MAX_LENGTH = std::string("mul(123,123)").length();

bool section_length_valid(std::string_view section) {
  size_t length = section.length();
  return length <= 3 && length >= 1;
}

std::optional<int> get_value(std::string_view view) {
  if (std::string(view.substr(0, START_LENGTH)) != "mul(") {
    return std::nullopt;
  }

  std::string_view numerical_section =
      view.substr(START_LENGTH, MAX_LENGTH - START_LENGTH);
  std::vector<int> numbers;
  size_t start = 0;

  for (size_t i = 0; i < numerical_section.length(); ++i) {
    char current_char = numerical_section[i];

    if (current_char == ',') {
      std::string_view section = numerical_section.substr(start, i - start);
      if (!section_length_valid(section) || numbers.size() != 0) {
        return std::nullopt;
      }
      numbers.push_back(std::stoi(std::string(section)));
      start = i + 1;
    } else if (current_char == ')') {
      std::string_view section = numerical_section.substr(start, i - start);

      if (!section_length_valid(section) || numbers.size() != 1) {
        return std::nullopt;
      }
      numbers.push_back(std::stoi(std::string(section)));
      break;
    } else {
      if (!std::isdigit(current_char)) {
        return std::nullopt;
      }
    }
  }

  if (numbers.size() != 2) {
    return std::nullopt;
  }

  return numbers[0] * numbers[1];
}

void three_a() {
  std::ifstream file = getMainFile(3);

  int sum = 0;
  std::string line;
  while (std::getline(file, line)) {
    std::string_view view(line);

    for (size_t i = 0; i < view.length(); ++i) {
      if (line[i] == 'm') {
        auto result = get_value(view.substr(i, MAX_LENGTH));
        if (result.has_value()) {
          sum += result.value();
        }
      }
    }
  }

  std::cout << sum << "\n";
}

const std::string do_string = std::string("do()");
const std::string dont_string = std::string("don't()");

std::optional<bool> change_state(std::string_view view) {
  if (std::string(view.substr(0, do_string.length())) == do_string) {
    return true;
  }

  if (std::string(view.substr(0, dont_string.length())) == dont_string) {
    return false;
  }

  return std::nullopt;
}

void three_b() {
  std::ifstream file = getMainFile(3);

  int sum = 0;
  bool enabled = true;
  std::string line;
  while (std::getline(file, line)) {
    std::string_view view(line);

    for (size_t i = 0; i < view.length(); ++i) {
      if (line[i] == 'm') {
        if (!enabled) {
          continue;
        }

        auto result = get_value(view.substr(i, MAX_LENGTH));
        if (result.has_value()) {
          sum += result.value();
        }
      }

      if (line[i] == 'd') {
        auto result = change_state(view.substr(i, dont_string.length()));

        if (result.has_value()) {
          enabled = result.value();
        }
      }
    }
  }

  std::cout << sum << "\n";
}

void three() { three_b(); }

File: 3/3.py
import sys
import os
from typing import List

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from utils.operate_on_lines import operate_on_lines

MAX_LENGTH = len("mul(123,123)")


def section_length_valid(section: str):
    length = len(section)
    return length <= 3 and len(section) >= 1


def get_value(string: str):
    if not string.startswith("mul("):
        return None

    numerical_section = string[len("mul(") : MAX_LENGTH]

    numbers: List[int] = []
    start = 0
    for i, char in enumerate(numerical_section):
        if char == ",":
            section = numerical_section[start:i]
            if not section_length_valid(section) or len(numbers) != 0:
                return None
            numbers.append(int(section))
            start = i + 1
        elif char == ")":
            section = numerical_section[start:i]
            if not section_length_valid(section) or len(numbers) != 1:
                return None
            numbers.append(int(section))
            break
        else:
            current_char_str = str(char)
            if not current_char_str.isdigit():
                return None

    if len(numbers) != 2:
        print(numbers)
        return None
    return numbers[0] * numbers[1]


def three_a():
    total_sum = [0]

    def fn(stripped_line: str):
        for i in range(len(stripped_line)):
            if stripped_line[i] == "m":
                value = get_value(stripped_line[i : i + MAX_LENGTH])
                if value is not None:
                    total_sum[0] += value

    operate_on_lines(fn)

    print(total_sum[0])


def change_state(string: str):
    if string.startswith("do()"):
        return True

    if string.startswith("don't()"):
        return False

    return None


class State:
    def __init__(self):
        self.sum = 0
        self.enabled = True


def three_b():
    state = State()

    def fn(stripped_line: str):
        for i in range(len(stripped_line)):
            if stripped_line[i] == "m":
                if not state.enabled:
                    continue

                value = get_value(stripped_line[i : i + MAX_LENGTH])
                if value is not None:
                    state.sum += value

            if stripped_line[i] == "d":
                new_enabled_state = change_state(stripped_line[i : i + len("don't()")])

                if new_enabled_state is not None:
                    state.enabled = new_enabled_state

    operate_on_lines(fn)
    print(state.sum)


if __name__ == "__main__":
    three_b()
